### PR TITLE
Adds DicomFile class with method for filtering tags

### PR DESCRIPTION
**Pull request by cppcooper**
Adds a bit of code to register an incoming dicom instance filter using `FilterCallback` as the callback function. Also adds a blank `Anonymize` function with comments explaining what to implement later.

### CODE REVIEW

**Comment by iwanlevin** [comment]
We should try to implement an exception that we can throw for the error on `FilterCallback` instead of just returning -1

**Comment by cppcooper** [comment]
> We should try to implement an exception that we can throw for the error on `FilterCallback` instead of just returning -1

I disagree, because the [filter callback is supposed to return an error code](https://sdk.orthanc-server.com/group__Toolbox.html#ga40738a61e1b77ed4945347b61f4b5483). So we should presume that throwing an exception will not be caught by the caller of the filter callback. If it is we have no way of telling where or how safely the server can continue. Besides both those reasons, we shouldn't even experience server crashing errors (which might require exceptions). If we identify one then we should log an error, and potentially throw an exception if it can't be recovered from any other way.

Exception handling is expensive, so C/C++ both have a tendency of using error prevention over handling and cheap handling over exception handling. I'm sure there is some modern C++ programmer who'd argue with me, but this exemplifies my experience.

File: src/plugin.cpp [hunk 1]
```diff
@@ -33,4 +46,43 @@ extern "C" {
...
+}
+
+void PopulateFilterList(){
+    nlm::json config(OrthancPluginGetConfiguration(context));
+
+    for(const auto &iter : config["Dicom-Filter"]["tags"]){
+        auto tag = iter.get<std::string>();
+        std::string temp = tag.substr(0,4);
+        tag.erase(0,5);
+        tag.append(temp);
```
Comment (cppcooper) [inline src/plugin.cpp:68]: Could the `substr` be inlined with the `append`? then perform the erase after?

**Comment by cppcooper** [comment]
binary parsing was worked out here: https://replit.com/@cppcooper/dicom-tags#main.cpp

**Comment by cppcooper** [comment]
```
orthanc-server  | E1107 12:39:43.419896 PluginsManager.cpp:164] Parse dicom data
orthanc-server  | E1107 12:39:43.419910 PluginsManager.cpp:164] Filter: read tag (2,0)=>2
orthanc-server  | E1107 12:39:43.419920 PluginsManager.cpp:164] Filter: read tag (2,1)=>65538
orthanc-server  | E1107 12:39:43.419926 PluginsManager.cpp:164] Filter: read tag (2,2)=>131074
orthanc-server  | E1107 12:39:43.419932 PluginsManager.cpp:164] Filter: read tag (2,3)=>196610
orthanc-server  | E1107 12:39:43.419938 PluginsManager.cpp:164] Filter: read tag (2,10)=>1048578
orthanc-server  | E1107 12:39:43.419943 PluginsManager.cpp:164] Filter: read tag (2,12)=>1179650
orthanc-server  | E1107 12:39:43.419948 PluginsManager.cpp:164] Filter: read tag (8,8)=>524296
orthanc-server  | E1107 12:39:43.419953 PluginsManager.cpp:164] Filter: read tag (8,16)=>1441800
orthanc-server  | E1107 12:39:43.419959 PluginsManager.cpp:164] Filter: read tag (8,18)=>1572872
orthanc-server  | E1107 12:39:43.419965 PluginsManager.cpp:164] Filter: read tag (8,20)=>2097160
orthanc-server  | E1107 12:39:43.419973 PluginsManager.cpp:164] Filter: read tag (8,30)=>3145736
orthanc-server  | E1107 12:39:43.419977 PluginsManager.cpp:164] Filter: read tag (8,50)=>5242888
orthanc-server  | E1107 12:39:43.419982 PluginsManager.cpp:164] Filter: read tag (8,60)=>6291464
orthanc-server  | E1107 12:39:43.419987 PluginsManager.cpp:164] Filter: read tag (8,70)=>7340040
orthanc-server  | E1107 12:39:43.419992 PluginsManager.cpp:164] Filter: read tag (8,80)=>8388616
orthanc-server  | E1107 12:39:43.419997 PluginsManager.cpp:164] Filter: read tag (8,81)=>8454152
orthanc-server  | E1107 12:39:43.420001 PluginsManager.cpp:164] Filter: read tag (8,90)=>9437192
orthanc-server  | E1107 12:39:43.420007 PluginsManager.cpp:164] Filter: read tag (8,1030)=>271581192
orthanc-server  | E1107 12:39:43.420013 PluginsManager.cpp:164] Filter: read tag (8,1050)=>273678344
orthanc-server  | E1107 12:39:43.420018 PluginsManager.cpp:164] Filter: read tag (8,2110)=>554696712
orthanc-server  | E1107 12:39:43.420023 PluginsManager.cpp:164] Filter: read tag (8,2112)=>554827784
orthanc-server  | E1107 12:39:43.420026 PluginsManager.cpp:164] Filter: nothing to do
```
I tested on docker, and as it turns out the parser does not get all of the data elements of the dicom file. I think because (based on the data for the elements where it stops) the dicom file nests tags under (8,2112).

Gonna need to return the the replit test code.

**Comment by cppcooper** [comment]
Okay, so the filter "works"
```
orthanc-server  | E1108 16:12:41.603301 PluginsManager.cpp:164] Filter: receiving dicom
orthanc-server  | E1108 16:12:41.603322 PluginsManager.cpp:164] Parse dicom data
orthanc-server  | E1108 16:12:41.603497 PluginsManager.cpp:164] filtering tag: 0010,0010
orthanc-server  | E1108 16:12:41.603514 PluginsManager.cpp:164] filtering tag: 0010,0020
orthanc-server  | E1108 16:12:41.603525 PluginsManager.cpp:164] filtering tag: 0010,0030
orthanc-server  | E1108 16:12:41.603537 PluginsManager.cpp:164] filtering tag: 0010,0040
orthanc-server  | E1108 16:12:41.604204 PluginsManager.cpp:164] Filter: compile new dicom
orthanc-server  | E1108 16:12:41.604781 PluginsManager.cpp:164] Filter: saving filtered dicom
orthanc-server  | E1108 16:12:41.605329 PluginsManager.cpp:164] Filter: cleanup
```
It does not, as it seems.. save anything to disk (I think, if it does I don't know where; I only checked the orthanc mount directory)

File: src/dicom-filter.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    const char* buffer = new char[new_size];
+    OrthancPluginLogWarning(globals::context, "Filter: compile new dicom buffer");
+    for(auto pair : keep_list){
+        size_t copy_size = pair.second-pair.first;
+        memcpy((void*)(buffer+i), (void*)(readable_buffer+pair.first), copy_size);
+        i += copy_size;
+    }
+    OrthancPluginLogWarning(globals::context, "Filter: create new dicom instance");
+    // todo: save dicom instance, cause apparently CreateDicomInstance doesn't
+    return OrthancPluginCreateDicomInstance(globals::context, buffer, new_size);
```
Comment (cppcooper) [inline src/dicom-filter.cpp:90]: Probably leaks memory

**Pull request merged**